src/ir/module-utils.cpp

```diff
@@ -136,8 +168,24 @@ DataSegment* copyDataSegment(const DataSegment* segment, Module& out) {
 // Copies named toplevel module items (things of kind ModuleItemKind). See
 // copyModule() for something that also copies exports, the start function, etc.
 void copyModuleItems(const Module& in, Module& out) {
+  std::vector<Index> fileIndexMap;
+  std::unordered_map<std::string, Index> debugInfoFileIndices;
+  for (Index i = 0; i < out.debugInfoFileNames.size(); i++) {
+    debugInfoFileIndices[out.debugInfoFileNames[i]] = i;
+  }
+  for (Index i = 0; i < in.debugInfoFileNames.size(); i++) {
+    std::string file = in.debugInfoFileNames[i];
+    auto iter = debugInfoFileIndices.find(file);
+    if (iter == debugInfoFileIndices.end()) {
+      Index index = out.debugInfoFileNames.size();
+      out.debugInfoFileNames.push_back(file);
+      debugInfoFileIndices[file] = index;
+    }
+    fileIndexMap.push_back(debugInfoFileIndices[file]);
+  }
+
   for (auto& curr : in.functions) {
-    copyFunction(curr.get(), out);
+    copyFunction(curr.get(), out, Name(), fileIndexMap);
```

Comment:
For efficiency, how about only passing a non-nullopt `fileIndexMap` if `!in.debugInfoFileNames.empty()` - ? That is, when the input has no debug info, we can avoid this work.

(`fileIndexMap` can be a `std::optional` that is only set if we have work to do, so when we don't it'll be nullopt.)

Reply:
I have also added a test at the beginning of `debug::copyDebugInfo` to just return if there is nothing to copy. This seems more foolproof than putting the test outside of the function...